client/rpc.go

```diff
@@ -4,70 +4,260 @@ package client
+		subscription *Subscription
+	}
+
+	// Subscription is returned by Subscribe() and can be used to iterate
+	// over the deposit and balance proofs.
+	Subscription struct {
+		perunlog.Embedding
+		// deposit proofs from the OP will be written into this channel.
+		depProofs chan tee.DepositProof
+		// balance proofs from the OP will be written into this channel.
+		balProofs chan tee.BalanceProof
+	}
+
+	// callback is used for handling call results.
+	callback func(wire.Result, []byte)
+)
 
-func NewRPC(host string, port uint16) *RPC {
-	conn, err := rpc.DialHTTP("tcp", fmt.Sprintf("%s:%d", host, port))
+// NewRPC returns a new RPC object.
+// RPC immediately tries to connect to the operator and starts to handle
+// incomming data.
+// You may want to call Subscribe afterwards if you need balance and/or
+// deposit proofs.
+func NewRPC(host string, port uint16) (*RPC, error) {
+	u := url.URL{Scheme: "ws", Host: fmt.Sprintf("%s:%d", host, port), Path: "/ws"}
+	conn, _, err := gorilla.DefaultDialer.Dial(u.String(), nil)
 	if err != nil {
-		panic(err)
+		return nil, err
 	}
-	return &RPC{conn}
+	rpc := &RPC{
+		conn:      conn,
+		callbacks: make(map[wire.ID]callback),
+		Embedding: perunlog.MakeEmbedding(perunlog.WithField("role", "client")),
+	}
+	if err := conn.SetReadDeadline(time.Time{}); err != nil {
+		return nil, err
+	}
+	conn.SetCloseHandler(func(code int, err string) error {
+		rpc.Log().WithField("code", code).Errorf("WS connection was closed: %s", err)
+		return rpc.Close()
+	})
+	go rpc.handleConnections()
+
+	return rpc, nil
 }
 
-func (r *RPC) AddTX(ctx context.Context, tx tee.Transaction) error {
-	var _reply struct{}
-	return r.call(ctx, "RemoteEnclave.AddTransaction", tx, &_reply)
+// SendTx sends one transaction to the operator.
+func (r *RPC) SendTx(ctx context.Context, tx tee.Transaction) error {
+	call := wire.NewSendTx(r.nextID(), tx)
+	errChan := make(chan error)
+	// Setup async response cb.
+	r.registerCallback(call.Call.ID, func(result wire.Result, msg []byte) {
+		if result.Error != "" {
+			errChan <- fmt.Errorf("SendTx RPC result: %s", result.Error)
+		} else {
+			errChan <- nil
+		}
+	})
+	// Make the call.
+	if err := r.sendJSON(call); err != nil {
+		return fmt.Errorf("sending json object: %w", err)
+	}
+	// Return error from async response cb.
+	select {
+	case err := <-errChan:
+		return err
+	case <-ctx.Done():
+		return ctx.Err()
+	}
 }
 
-func (r *RPC) GetDepositProof(ctx context.Context, epoch uint64, user common.Address) (reply tee.DepositProof, err error) {
-	for i := 0; ; i++ {
-		if e := r.call(ctx, "RemoteEnclave.GetDepositProof", user, &reply); e != nil {
-			return tee.DepositProof{}, e
+// Subscribe subscribes to the Balance and Deposit proof topic.
+// The user must always read the proofs from via `DepositProof` and
+// `BalanceProof`.
+func (r *RPC) Subscribe(ctx context.Context, user common.Address) (*Subscription, error) {
+	r.subMtx.Lock()
+	r.subscription = &Subscription{
```

Comment:
Do we even need the `subMtx`? It looks to me that `RPC.Subscribe` is called once during setup of the client and from then on is used, but never reset, no? Is it even valid to call `Subscribe` twice?

Reply:
The client assumes it to be valid to subscribe twice, because otherwise that case could not be tested.  
The RPC Server rejects a double subscription obviusly.

Reply:
But i guess since we only call it twice in testing code, i will remove it.